### PR TITLE
feat(sbb-container): first implementation

### LINKED ISSUE #1908: story(sbb-container, sbb-sticky-bar): initial implementation
We want to implement a container component and associated sticky bar.
The container component has the same spacing behavior as the header and footer (including the `expanded` property).

[Figma Spec](https://www.figma.com/file/9r6xSfNmEfCFxl1yFYedrj/Lyne-Components?type=design&node-id=25018%3A468481&mode=design&t=MDGURDdCkcBSQrfL-1)

**References**

- https://www.figma.com/proto/HeBonE3QY13ABQ4f2FShTJ/lyne-projects?page-id=0%3A1&type=design&node-id=2-182675&viewport=1321%2C1370%2C0.32&t=Kaic1GfC9eEVmjNn-1&scaling=min-zoom

**Design/Technical  Spec**

`<sbb-container>`

- Have the same spacing as `<sbb-header>` and `<sbb-footer>`, including a `expanded` property
- Use the sass mixin for spacing as is used in header and footer

`<sbb-sticky-bar>`

- Can be slotted into `<sbb-container>` and assumes its horizontal spacing
- Should stick to the bottom, while the container is visible
- Should have a upper shadow when it is sticky at the bottom
- Logic for the shadow should be self-contained in the `<sbb-sticky-bar>` (if possible)
- Sticky at the top is out-of-scope at the moment

```html
<sbb-container>
  my content
</sbb-container>

// sticky bar
<sbb-container>
  my content
  <sbb-sticky-bar>
    sticky content
  </sbb-sticky-bar>
</sbb-container>
```

---

_Original request_

### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/lyne-design-system/lyne-components/blob/master/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/lyne-design-system/lyne-components/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://github.com/orgs/lyne-design-system/projects/4) for an issue that matches the one I want to file, without success.

### Request type

Request for a new component

### It affects the following packages

Components

### Functionality

The bar should be created including the transition of the stickyness into the content. The content of the sticky bar is a placeholder. It is important that the spacing below the placeholder in the embedded state is subtracted from the section spacing that follows.

### Link to design proposal file

https://www.figma.com/file/9r6xSfNmEfCFxl1yFYedrj/Lyne-Components?type=design&node-id=25018%3A468482&mode=design&t=t3CgF6sOkvbr2UNz-1

### PR DESCRIPTION
**Pull request by MarioCastigliano**
## Preflight Checklist

- [x] I have read the [Contributing Guidelines](https://github.com/lyne-design-system/lyne-components/blob/master/CONTRIBUTING.md) for this project.
- [x] I agree to follow the [Code of Conduct](https://github.com/lyne-design-system/lyne-components/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
- [x] I have searched the [pull request tracker](https://github.com/lyne-design-system/lyne-components/pulls) for a Pull Request (PR) that matches the one I want to submit, without success.

## Issue

This PR Closes #1908 

## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)

See [Review Guidelines](../REVIEW.md) for more information what is checked during review process.

## Changes

Changes in this pull request:

-

## Browsers

I tested the build on the following browsers:

- [x] Firefox Desktop
- [x] Chrome Desktop
- [x] Edge Desktop
- [x] Safari Desktop
- [x] Chrome Mobile
- [ ] Safari Mobile

## Screen readers

I tested the build on the following browsers:

- [ ] JAWS Firefox Desktop
- [ ] JAWS Chrome Desktop
- [ ] NVDA Firefox Desktop
- [ ] NVDA Chrome Desktop
- [x] VoiceOver Safari Desktop
- [x] VoiceOver Chrome Desktop
- [ ] VoiceOver Safari Mobile
- [x] Android Accessibility Suite Chrome Mobile

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/components/container/container/container.spec.ts
Comment (DavideMininni-Fincons) [inline src/components/container/container/container.spec.ts:3]: This could be removed

File: src/components/container/container/container.stories.ts
Comment (DavideMininni-Fincons) [inline src/components/container/container/container.stories.ts:27]: `styleMap` can be removed here, since there's no variable interpolation

File: src/components/container/container/container.stories.ts
Comment (DavideMininni-Fincons) [inline src/components/container/container/container.stories.ts:60]: Use `variant.options[0]`

File: src/components/container/sticky-bar/readme.md
Comment (DavideMininni-Fincons) [inline src/components/container/sticky-bar/readme.md:1]: Don't forget to write docs before changing to 'ready to review' :)

File: src/components/container/sticky-bar/sticky-bar.stories.ts
Comment (DavideMininni-Fincons) [inline src/components/container/sticky-bar/sticky-bar.stories.ts:37]: See other comments about `styleMap`

File: src/components/container/sticky-bar/sticky-bar.stories.ts
Comment (DavideMininni-Fincons) [inline src/components/container/sticky-bar/sticky-bar.stories.ts:52]: Same as others (`stylemap`)

**Review by DavideMininni-Fincons — COMMENTED** [review]
Good work, I added some comments ;)

File: src/components/container/container/container.ts
Comment (TomMenga) [inline src/components/container/container/container.ts:77]: Since Lit, we can avoid querying inside the shadow dom of other components.
I'd prefer the sticky bar to have an `@internal` public getter that exposes the element you need

File: src/components/container/container/container.ts
Comment (TomMenga) [inline src/components/container/container/container.ts:33]: Discussion point:
Shouldn't the sticky bar animation be the responsibility of the sticky bar component?

File: src/components/container/sticky-bar/sticky-bar.scss
Comment (TomMenga) [inline src/components/container/sticky-bar/sticky-bar.scss:27]: Discussion point:
I understand why `width: 100vw;` is needed but it prevents any possible use of a non-full page sticky bar (imagine a page where you have a side menu for example).

The best solution would be to negate the spacing of the `sbb.page-spacing; | sbb.page-spacing-expanded` but I understand this might be hard to achieve.
A good compromise could be to a CSS variable defined on the `:host` so that the consumer  has a way to handle those scenarios

File: src/components/container/container/container.spec.ts
Comment (TomMenga) [inline src/components/container/container/container.spec.ts:11]: Just to be consistent with the sticky-bar test, we can avoid using `equalSnapshot()` since it is a very simple template

**Review by TomMenga — DISMISSED** [review]
Good job. 👍 
Some small changes and some topic to discuss

Ps. Don't forget to approve the new stories on chromatic when you get the ok from the UX

File: src/components/container/sticky-bar/sticky-bar.scss
Comment (MarioCastigliano) [reply src/components/container/sticky-bar/sticky-bar.scss:27]: We asked Manuel about this and he was positive the sticky bar is only meant to be used in full width

Comment (TomMenga) [reply src/components/container/sticky-bar/sticky-bar.scss:27]: In any case, I think it is always better to anticipate the changes

Comment (MarioCastigliano) [reply src/components/container/sticky-bar/sticky-bar.scss:27]: After discussing the matter with @kyubisation, we decided to try with an "infinite" sticky-bar with `overflow-x:hidden` on the sbb-container. 
But after some tests I noticed that any kind of overflow rule prevents the `position: sticky` from working. Another solution could be `contain: contents` (which works just like `overflow: hidden`), which does not cause any problems with the sticky positioning, but it messes up any overlay that goes outside the container. 
So we'll use the css variable approach

File: src/components/container/sticky-bar/sticky-bar.scss
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.scss:17]: I asked Florian if we really should provide a default padding. I remember we once decided to not include any default padding.

File: src/components/container/sticky-bar/sticky-bar.scss
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.scss:30]:
```suggestion
@include sbb.absolute-center-x
```

Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.scss:30]:
```suggestion
@include sbb.absolute-center-x
```

File: src/components/container/sticky-bar/sticky-bar.scss
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.scss:32]: 1s seems long for an animation?

But first a general question, would it not just work if we detect sticky state of sticky bar and just us transition instead of complicated delayed animation? I remember I once implemented such a sticky bar in another project and also privately (www.lorenapeier.ch If you like to test it :) ). basically I just used an intersection observer. something like discussed here: https://stackoverflow.com/questions/16302483/event-to-detect-when-positionsticky-is-triggered 
This would save you from using scroll listener.

this slowly (depending on scroll position) fade in animation causes also bad situations where the shadow should be visible but is only very slightly seeable:
Bildschirmfoto 2023-12-20 um 10 37 01

File: src/components/container/sticky-bar/sticky-bar.scss
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.scss:36]: it's not wrong but normally we use data attributes to describe state of a component instead of classes.

File: src/components/container/sticky-bar/sticky-bar.scss
Comment (jeripeierSBB) [reply src/components/container/sticky-bar/sticky-bar.scss:32]: Other than that, when the sticky bar is not sticky anymore, the shadow is still visible (seen on https://lyne-components-preview-pr2271.app.sbb.ch/?path=/story/components-sbb-container-sbb-sticky-bar--with-content-after) which is somehow working in fullscreen but not in stories view. other than that in fullscreen the container below doesn't have a background color but in inline view it has.
Bildschirmfoto 2023-12-20 um 10 39 22

and sometimes the background color is transparent even if sticky:
Bildschirmfoto 2023-12-20 um 10 48 50

File: src/components/container/sticky-bar/sticky-bar.stories.ts
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.stories.ts:53]:
```suggestion
```

File: src/components/container/sticky-bar/sticky-bar.stories.ts
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.stories.ts:63]:
```suggestion
  layout: 'fullscreen',
  title: 'components/sbb-container/sbb-sticky-bar',
```

File: src/components/container/sticky-bar/sticky-bar.stories.ts
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.stories.ts:41]: Please add some color variants here as well to show inheriting color of container

File: src/components/container/container/container.ts
Comment (jeripeierSBB) [inline src/components/container/container/container.ts:19]:
```suggestion
  @property({ reflect: true }) public color: 'transparent' | 'white' | 'milk' | 'midnight' =
```

**Review by jeripeierSBB — CHANGES_REQUESTED** [review]
Thanks for the implementing these components! There are some things to address and also topics to discuss probably.

- I think we should use the sbb-container in our stories where possible, so at least in the home stories (instead of the grid, at least when container supports grid :) ).
- In page layout docs (src/storybook/styles/layout/readme.md) we should probably mention the new possibility of the sbb-container.

File: src/components/container/container/container.scss
Comment (MarioCastigliano) [reply src/components/container/container/container.scss:11]: In the transparent version they don't have the same color. As the container is actually set to transparent but the sticky-bar needs to be white

Comment (jeripeierSBB) [reply src/components/container/container/container.scss:11]: yes but here on container level it's always the same, so I'd still argue for the change :)

File: src/components/container/container/container.ts
Comment (jeripeierSBB) [inline src/components/container/container/container.ts:48]: on slotchange, the sticky bar should be updated with expanded as well. You can at anytime insert or remove a sticky bar into the container.

Comment (jeripeierSBB) [inline src/components/container/container/container.ts:48]: on slotchange, the sticky bar should be updated with expanded as well. You can at anytime insert or remove a sticky bar into the container.

File: src/components/container/sticky-bar/sticky-bar.stories.ts
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.stories.ts:79]: why not just passing color into containerContent and only once check for color === 'midnight'?

File: src/components/container/container/container.ts
Comment (jeripeierSBB) [reply src/components/container/container/container.ts:48]: alternative approach is to check for parent's expanded in connectedcallback of the sticky bar.

File: src/components/container/sticky-bar/sticky-bar.stories.ts
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.stories.ts:141]:
```suggestion
```

File: src/components/container/sticky-bar/sticky-bar.ts
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.ts:57]: As it's made as block element in scss, we could directly use a div? I'd prefer to have a class name which describes the purpose of the div.
```suggestion
      <div class="sbb-sticky-bar__intersector"></div>
```

File: src/components/container/sticky-bar/sticky-bar.ts
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.ts:37]:
```suggestion
      this._intersector = this.shadowRoot.querySelector('.sbb-sticky-bar__intersector');
```

File: src/components/container/sticky-bar/sticky-bar.ts
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.ts:41]: can never be different
```suggestion
    this._observer.observe(this._intersector);
```

File: src/components/container/sticky-bar/sticky-bar.ts
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.ts:34]: The only place where this function is called is firstUpdated(), so do we really need the disconnect and reconnection? I feel that it could be necessary to call it in connectedcallback as well. Can you test removing, moving and re adding a sticky bar in DOM?

**Review by jeripeierSBB — CHANGES_REQUESTED** [review]
I like the new approach! There are some topics which needed to be addressed.

File: src/components/container/container/container.scss
Comment (jeripeierSBB) [reply src/components/container/container/container.scss:40]: discussed with @kyubisation, we don't create a property, we don't include grid in container for now.

File: src/components/container/container/container.stories.ts
Comment (jeripeierSBB) [inline src/components/container/container/container.stories.ts:8]:
```suggestion
```

File: src/components/container/sticky-bar/sticky-bar.stories.ts
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.stories.ts:76]: Maybe nest in an action group? What would be the best example here? @mcilurzo

File: src/components/container/sticky-bar/sticky-bar.scss
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.scss:20]: transition-timing-function missing. maybe try with var(--sbb-animation-easing) and introduce --sbb-sticky-bar-animation-easing var

File: src/components/container/sticky-bar/sticky-bar.scss
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.scss:16]: It should not really change something, but this is clearly a block element for me :)
```suggestion
  display: block;
```

**Review by jeripeierSBB — CHANGES_REQUESTED** [review]
Some few additional topics left:
- Possibility to define separate color for sticky bar when sticky.
- High Contrast Mode support
- In page layout docs (src/storybook/styles/layout/readme.md) we should mention the new possibility of the sbb-container.

File: src/components/container/sticky-bar/sticky-bar.stories.ts
Comment (mcilurzo) [reply src/components/container/sticky-bar/sticky-bar.stories.ts:76]: an action-group would be nice.

File: src/components/container/sticky-bar/sticky-bar.scss
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.scss]: I'd not toggle vars on .css class level as it's one level than then the host and is therefore "not the same".
Second question, should it not be transparent when not sticking? you could also directly write: "background-color: transparent", or then just toggling it on host level

File: src/components/container/sticky-bar/sticky-bar.scss
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.scss]: data-settled is not existing? 
```suggestion
:host([color='midnight']) {
```

File: src/components/container/sticky-bar/sticky-bar.scss
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.scss:81]: Please write a short comment why the 1 pixel and transformation is necessary. Is it worth to have a css var for the intersector height, or could it be inlined as an internal detail?

File: src/components/container/sticky-bar/sticky-bar.stories.ts
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.stories.ts:78]: Same question, needed and why?

**Review by jeripeierSBB — CHANGES_REQUESTED** [review]

File: src/components/container/sticky-bar/sticky-bar.scss
Comment (MarioCastigliano) [reply src/components/container/sticky-bar/sticky-bar.scss:81]: On windows the animation dind't work on any browser unless I gave the intersector an height. I'll document it with a comment. Inlined is fine to me.

File: src/components/container/container/container.stories.ts
Comment (MarioCastigliano) [reply src/components/container/container/container.stories.ts:14]: image

Otherwise when top element has a margin-top we get things like this

File: src/components/container/sticky-bar/sticky-bar.scss
Comment (jeripeierSBB) [reply src/components/container/sticky-bar/sticky-bar.scss]: It's definitely not that simple as I thought ;)

File: src/components/container/container/container.scss
Comment (jeripeierSBB) [inline src/components/container/container/container.scss:24]: Probably this creates a scrollbar if a consumer sets a fixed height on the container? With this I think we remove the possibility to let the overflow be controlled by consumer from outside. I'm not really understanding why margin in a slotted content does influence the parent container (here sbb-container). hmm. Will think about it :)

File: src/components/container/sticky-bar/sticky-bar.stories.ts
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.stories.ts:169]: for chromatic it would be great if we have two stories with a background color set in the container. one with color on sticky and one without. So we would have all combinations.

To test the sticking state in chromatic, maybe we can in some stories enforce a fixed height, so it shows sticky?

File: src/components/container/sticky-bar/sticky-bar.scss
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.scss:50]: We need the disableAnimation property to disable the animation

File: src/components/container/container/container.scss
Comment (jeripeierSBB) [reply src/components/container/container/container.scss:24]: I tested it, it's ok how it is now :) sorry..

File: src/components/container/sticky-bar/sticky-bar.ts
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.ts:44]:
```suggestion
```

File: src/components/container/sticky-bar/sticky-bar.scss
Comment (jeripeierSBB) [inline src/components/container/sticky-bar/sticky-bar.scss:36]:
```suggestion
:host([disable-animation]) {
```

File: src/components/container/sticky-bar/sticky-bar.stories.ts
Comment (jeripeierSBB) [reply src/components/container/sticky-bar/sticky-bar.stories.ts:211]: I see new stories but without fixed height, or is the sticky bar sticky in chromatic anyways?

**Review by jeripeierSBB — CHANGES_REQUESTED** [review]

**Review by jeripeierSBB — APPROVED** [review]
LGTM. Thanks for your endurance!

**Review by jeripeierSBB — APPROVED** [review]
LGTM

**Pull request merged**
